events/utils.py

```diff
@@ -38,3 +44,38 @@ def update(d, u):
         else:
             d[k] = u[k]
     return d
+
+
+def recache_n_events(keywords):
+    """
+    Recache the number of events for the given keywords.
+
+    :type keywords: Iterable[Keyword]
```

Comment:
I'd honestly suggest changing the signature of the function to `keyword_ids` (and obviously how the function works as well), so you never have to wrangle querysets or actual keyword objects within here, since you don't need them at all.

That way the function would be

```python
def recache_n_events(keyword_ids):
    """
    Recache the number of events for the given keywords (by ID).
    :type keywords: Iterable[str]
    """

    if not keywords:
        return

    with transaction.atomic():
        Keyword.objects.filter(id__in=keyword_ids).update(n_events=0)
        for keyword_id, n_events in count_events_for_keywords(keyword_ids).items():
            Keyword.objects.filter(id=keyword_id).update(n_events=n_events, n_events_changed=False)
```
and the caller would be free to pass in a `.values_list('id', flat=True)` or a set of keyword ids from a cache or whatever without having to worry about loading the full objects -- and if they _do_ have keyword objects, they can pass in `(k.id for k in keyword)`.